Always use shortest serialization for columns CSS property

nt1m opened the pull request:
#### 492c75edf7c72accb07482a017d6e42e9f357dde
<pre>
Always use shortest serialization for columns CSS property
<a href="https://bugs.webkit.org/show_bug.cgi?id=247826">https://bugs.webkit.org/show_bug.cgi?id=247826</a>
rdar://102287297

Reviewed by Darin Adler.

Always omit `auto` if 2 values are specified and one of them is `auto`. In the case of `auto auto`, keep one `auto` for serialization.

* LayoutTests/fast/multicol/column-width-zero-expected.txt: Removed.
* LayoutTests/fast/multicol/column-width-zero.html: Removed.
Remove conflicting WebKit tests, covered by WPT.

* LayoutTests/fast/css/getPropertyValue-columns-expected.txt:
* LayoutTests/fast/css/getPropertyValue-columns.html:
* LayoutTests/fast/multicol/columns-shorthand-parsing-2-expected.txt:
* LayoutTests/fast/multicol/columns-shorthand-parsing-2.html:
Update WebKit test result expectations.

* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/zero-column-width-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-width-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-width-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-valid.html:
Rebaseline newly passing WPT and add more coverage.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeColumnWidth): Allow unitless zero per-spec, there is no ambiguity in the shorthand as the minimum value for column-count is 1.
(WebCore::CSSPropertyParser::consumeColumns):

Canonical link: <a href="https://commits.webkit.org/256625@main">https://commits.webkit.org/256625@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edefd43963dc0a4e02cf2adbb9508ef8ee581dc6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96384 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5638 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105927 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166271 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100367 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5805 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34384 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88763 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102653 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102055 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82984 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31304 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88051 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74187 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40114 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37788 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20932 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4598 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/2906 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/607 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40201 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/fast/multicol/columns-shorthand-parsing-2-expected.txt:
```diff
@@ -3,49 +3,56 @@ Test the behavior when 'auto' is part of the 'columns' property value. See http:
 
 
-PASS element.style.columnWidth is "auto"
+FAIL element.style.columnWidth should be auto. Was initial.
```
What’s the deal with these? Is the test broken or the code? If it’s the test, can we fix the test?

darinadler replied:
I think these failures are legit. We should not be setting the longhands to `initial`.

nt1m replied:
I agree, will fix that.

darinadler reviewed (approved):
The tests here cover fetching the specified style. Do we have tests that check the computed style as well? Does that already work correctly?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5536,44 +5532,32 @@ bool CSSPropertyParser::consumeColumns(bool important)
-    else
-        addProperty(CSSPropertyColumnCount, CSSPropertyInvalid, CSSValuePool::singleton().createIdentifierValue(CSSValueAuto), important, !hasPendingExplicitAuto /* implicit */);
-    
+    // If both values are auto, set column-width explicitly to auto so the shorthand serializes correctly.
+    if (!columnWidth && !columnCount)
+        columnWidth = CSSValuePool::singleton().createIdentifierValue(CSSValueAuto);
```
This behavior affects the columns shorthand, but also what the `column-width` and `column-count` properties return. Can we add a test that sets the `columns` to `auto` and then verifies how each of `column-count` and `column-width` serialize in that case? Is it correct that `column-width` serializes as `initial`?

nt1m replied:
This is what the failing subtests are about, but I'll add a more explicit one.

nt1m committed:
Always use shortest serialization for columns CSS property
https://bugs.webkit.org/show_bug.cgi?id=247826
rdar://102287297

Reviewed by Darin Adler.

Always omit `auto` if 2 values are specified and one of them is `auto`. In the case of `auto auto`, keep one `auto` for serialization.

* LayoutTests/fast/multicol/column-width-zero-expected.txt: Removed.
* LayoutTests/fast/multicol/column-width-zero.html: Removed.
Remove conflicting WebKit tests, covered by WPT.

* LayoutTests/fast/css/getPropertyValue-columns-expected.txt:
* LayoutTests/fast/css/getPropertyValue-columns.html:
* LayoutTests/fast/multicol/columns-shorthand-parsing-2-expected.txt:
* LayoutTests/fast/multicol/columns-shorthand-parsing-2.html:
Update WebKit test result expectations.

* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/zero-column-width-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-width-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-width-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/columns-valid.html:
Rebaseline newly passing WPT and add more coverage.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeColumnWidth): Allow unitless zero per-spec, there is no ambiguity in the shorthand as the minimum value for column-count is 1.
(WebCore::CSSPropertyParser::consumeColumns):

Canonical link: https://commits.webkit.org/256625@main
